app/grandchallenge/reader_studies/views.py

```diff
@@ -1323,8 +1317,126 @@ class DisplaySetDetail(
     raise_exception = True
 
 
+class InterfaceProcessingMixin:
+    def _process_new_interfaces(self, form):
+        new_interfaces = form.data.get("new_interfaces")
+        validated_data = {}
+        if new_interfaces:
+            errors = {}
+            for entry in new_interfaces:
+                interface = ComponentInterface.objects.get(
+                    pk=entry["interface"]
+                )
+                form = ComponentInterfaceCreateForm(
+                    data=entry,
+                    pk=None,
+                    interface=interface.pk,
+                    user=self.request.user,
+                    base_obj=self.reader_study,
+                    auto_id="%s",
+                    htmx_url=None,
+                )
+                if form.is_valid():
+                    cleaned = form.cleaned_data
+                    validated_data[cleaned["interface"].slug] = cleaned[
+                        interface.slug
+                    ]
+                else:
+                    errors.update(
+                        {entry["interface"]: form.errors[interface.slug]}
+                    )
+            if errors:
+                raise ValidationError(errors)
+        return validated_data
+
+    def return_errors(self, errors):
+        return JsonResponse(errors, status=400)
+
+    def form_invalid(self, form):
+        errors = form.errors
+        try:
+            self._process_new_interfaces(form)
+        except ValidationError as e:
+            errors.update(e.message_dict)
+        return self.return_errors(errors)
+
+    def form_valid(self, form):
+        errors = {}
+        data = {}
+        try:
+            data = self._process_new_interfaces(form)
+        except ValidationError as e:
+            errors.update(e.message_dict)
+
+        try:
+            data.update(form.cleaned_data)
+            self.process_data_for_object(data)
+        except ValidationError as e:
+            errors.update(e.message_dict)
+
+        if errors:
+            return self.return_errors(errors)
+        else:
+            messages.success(
+                self.request,
+                "Display set created. Image and file import jobs have been queued.",
+            )
+            return JsonResponse({"redirect": self.get_success_url()})
```

Comment:
This is again a bit awkward: we're returning errors on form_valid. I took this over as is because I can't think of a different way of doing this. The reason we do this is that we add new interfaces to a display set and they can't be validated as part of the normal form.

Reply:
Without looking at this in detail shouldn't it be done in a custom `clean` method in a form?

Reply:
Yes, I'm actually just working on that! ;-)